### DISCUSSION TITLE
My LSTM model is stuck on a constant value

### DISCUSSION THREAD

**Discussion by Ystyphn** [mrdbourke/pytorch-deep-learning#1350] [Q&A]
# I have an LSTM model: Problem is the evaluation returns stuck values.

Context: I am running on Jupyter Lab

First this is the Dataset that I made.
```
from torch.utils.data import Dataset, DataLoader
from typing import Tuple

class ForexDatasetV0(Dataset):
    def __init__(self, input_tensor: torch.Tensor, expected_tensor: torch.Tensor):
        if type(input_tensor) != torch.Tensor or type(expected_tensor) != torch.Tensor:
            raise TypeError(f"Input and expected should be of type tensor. Got: {type(input_tensor)} and {type(expected_tensor)} instead")
        if len(input_tensor) != len(expected_tensor):
            raise Exception(f"Input and expected tensor should be of the same length")
            
        self.input_tensor: torch.Tensor = input_tensor
        self.expected_tensor: torch.Tensor = expected_tensor


    def __len__(self):
        return len(self.input_tensor)


    def __getitem__(self, idx) -> Tuple[torch.Tensor, torch.Tensor]:
        return self.input_tensor[idx], self.expected_tensor[idx]
```
Data loader and data setup

```
train_slice = round(len(inputs) * 0.8)
X_train, y_train = inputs_list[:train_slice], labels[:train_slice]
X_test, y_test = inputs_list[train_slice:], labels[train_slice:]

X_train, y_train = torch.from_numpy(X_train).type(torch.float32), torch.from_numpy(y_train).type(torch.float32)
X_test, y_test = torch.from_numpy(X_test).type(torch.float32), torch.from_numpy(y_test).type(torch.float32)

X_train.requires_grad_(True)
X_test.requires_grad_(True)

train_dataset = ForexDatasetV0(input_tensor=X_train, expected_tensor=y_train)
test_dataset = ForexDatasetV0(input_tensor=X_test, expected_tensor=y_test)

train_dataloader = DataLoader(train_dataset, batch_size=32, shuffle=False) # No need for shuffling because order is what I needed.
test_dataloader = DataLoader(test_dataset, batch_size=32, shuffle=False)
```

Now for the model:

```
class ForexModelV0(nn.Module):
    def __init__(self, input_size: int = 3, hidden_size: int = 10, output_size: int = 1, layers: int = 3):
        super().__init__()
        self.input_size: int = input_size
        self.hidden_size: int = hidden_size
        self.output_size: int = output_size
        self.layers: int = layers
        self.lstm = nn.LSTM(input_size=input_size, 
                            hidden_size=hidden_size, 
                            num_layers=layers, 
                            batch_first=True)
        self.final_linear = nn.Linear(self.hidden_size, self.output_size)


    def forward(self, x: torch.Tensor) -> torch.Tensor:
        #h_0 = torch.zeros(self.layers, x.shape[0], self.hidden_size)
        #c_0 = torch.zeros(self.layers, x.shape[0], self.hidden_size)

        out, _ = self.lstm(x)

        return self.final_linear(out[:, -1, :])

# Typically
model_0 = ForexModelV0()
```

**And these are the loss, accuracy and optimizer:**
```
from sklearn.metrics import mean_absolute_percentage_error

loss_fn = nn.HuberLoss()
optimizer = torch.optim.AdamW(params=model_0.parameters(), lr=0.1)
```

Then the training loop:

```
### Training models
def train_model(model: nn.Module, 
                data: DataLoader,
                optimizer_fn: torch.optim, 
                loss_fn: nn.Module, 
                acc_fn,
                epochs: int=3, 
                report_frequency: int=1 # How many epochs to run before  printing reports
               ):
    acc = 0.0
    ave_loss = 0.0
    loss = 0.0
    previous_data = 0.0
    
    model.train() # Yes, I also don't forget this.
    
    print("Is model training? ",model.training)
    for epoch in range(epochs):
        for X, y in tqdm(data):
            #if type(previous_data) is torch.Tensor: # This just me making sure that the data being inputted was different every run.
            #   print(torch.sum(torch.abs(previous_data - X))) # Turns out it was.
            prediction: torch.Tensor = model(X)
            loss = loss_fn(prediction, y.unsqueeze(1))
            ave_loss += loss
            p_clone = prediction.clone() # Why clone? I am just afraid that these might be passed by reference and mess things up...
            p_clone = p_clone.detach().numpy() # even more. So I make sure it was passed by value... I hope?
            acc = acc_fn(y_pred=p_clone, y_true=y) # Refers to the scikit learn MAPE (Mean Abs Percetage Error)
            optimizer_fn.zero_grad()
            loss.backward()
            optimizer_fn.step()
            previous_data = X
        if epoch % report_frequency == 0:
            print(f"\nEpoch: {epoch}")
            print("------------------------------")
            print(f"Average loss: {float(ave_loss/len(data))}")
            print(f"Mean accuracy percentage error: {float(acc):.2f}%")
            print(f"Shape of prediction and y_true: {p_clone.shape}, {y.unsqueeze(1).shape}")
            print("------------------------------")
        ave_acc = 0.0
        ave_loss = 0.0
```

Now after:
```
train_model(model=model_0, 
            data=train_dataloader, 
            optimizer_fn=optimizer, 
            loss_fn=loss_fn, 
            acc_fn=mean_absolute_percentage_error, 
            epochs=3, report_frequency=2)
```

No matter what the input, it is stuck on a constant output:
```
X_sample_1 = X_test[3]
X_sample_2 = X_test[5]

model_0.eval() # Yes, I don't forget this.
with torch.inference_mode():
    print(model_0(X_sample_1.unsqueeze(0)))
    print(model_0(X_sample_2.unsqueeze(0)))

# Output: 
# tensor([[0.0007]])
# tensor([[0.0007]])
# But it depends on the runtime. Sometimes it's -0.0007, sometimes stuck at 0.0003 and so on.
```
<img width="1462" height="262" alt="image" src="https://github.com/user-attachments/assets/16b928a8-7a6e-4bdf-986a-47d3539c22a8" />

Change it to 4, 6? Sure:
<img width="1468" height="241" alt="image" src="https://github.com/user-attachments/assets/7c393059-2077-4360-8b52-d21849b379f6" />
And yes, I do run the cell lol.

And no matter what I do, it's still stuck, much like me in this project. 

And yes, I do check the value of X_sample_1 and X_sample_2

```
tensor([[ 0.0000e+00, -5.6262e-05,  2.0132e-04],
        [-5.2340e-04, -1.0880e-04,  2.0662e-04],
        [ 1.0302e-04, -9.3790e-05,  2.1265e-04],
        [-3.0902e-04, -1.1310e-04,  2.1724e-04],
        [ 9.4453e-05, -1.1202e-04,  2.1823e-04],
        [ 2.5757e-04, -1.0074e-04,  2.3194e-04],
        [ 2.2317e-04, -9.2152e-05,  2.4106e-04],
        [-8.5817e-06, -8.5726e-05,  2.4183e-04],
        [-1.4589e-04, -6.0028e-05,  2.1015e-04],
        [ 6.0081e-05, -6.2169e-05,  2.0873e-04],
        [-8.5825e-06, -6.2705e-05,  2.0858e-04],
        [-2.5748e-05, -5.1453e-05,  2.0539e-04],
        [ 3.4331e-05, -4.0196e-05,  2.0485e-04],
        [-6.8660e-05, -3.8590e-05,  2.0452e-04],
        [-1.9741e-04, -2.6804e-05,  1.8901e-04],
        [-1.8887e-04, -4.3971e-05,  1.9047e-04],
        [-1.7173e-04, -5.4704e-05,  1.9251e-04],
        [ 1.8893e-04, -1.0183e-05,  1.5830e-04],
        [-2.8335e-04, -3.4331e-05,  1.6834e-04],
        [-9.7911e-04, -7.6211e-05,  2.7867e-04],
        [ 7.2216e-04, -3.6980e-05,  3.3784e-04],
        [-2.4914e-04, -6.8649e-05,  3.3245e-04],
        [ 8.5931e-05, -7.7227e-05,  3.2653e-04],
        [-3.6088e-04, -9.9245e-05,  3.3297e-04],
        [ 7.3061e-04, -4.4464e-05,  3.8830e-04],
        [-7.7302e-05, -5.3050e-05,  3.8741e-04],
        [-1.8039e-04, -6.3788e-05,  3.8840e-04],
        [-3.3506e-04, -8.3120e-05,  3.9369e-04],
        [-1.8048e-04, -9.6546e-05,  3.9312e-04],
        [-1.5472e-04, -1.0192e-04,  3.9329e-04]], grad_fn=<SelectBackward0>) 

 tensor([[ 1.0302e-04, -9.3790e-05,  2.1265e-04],
        [-3.0902e-04, -1.1310e-04,  2.1724e-04],
        [ 9.4453e-05, -1.1202e-04,  2.1823e-04],
        [ 2.5757e-04, -1.0074e-04,  2.3194e-04],
        [ 2.2317e-04, -9.2152e-05,  2.4106e-04],
        [-8.5817e-06, -8.5726e-05,  2.4183e-04],
        [-1.4589e-04, -6.0028e-05,  2.1015e-04],
        [ 6.0081e-05, -6.2169e-05,  2.0873e-04],
        [-8.5825e-06, -6.2705e-05,  2.0858e-04],
        [-2.5748e-05, -5.1453e-05,  2.0539e-04],
        [ 3.4331e-05, -4.0196e-05,  2.0485e-04],
        [-6.8660e-05, -3.8590e-05,  2.0452e-04],
        [-1.9741e-04, -2.6804e-05,  1.8901e-04],
        [-1.8887e-04, -4.3971e-05,  1.9047e-04],
        [-1.7173e-04, -5.4704e-05,  1.9251e-04],
        [ 1.8893e-04, -1.0183e-05,  1.5830e-04],
        [-2.8335e-04, -3.4331e-05,  1.6834e-04],
        [-9.7911e-04, -7.6211e-05,  2.7867e-04],
        [ 7.2216e-04, -3.6980e-05,  3.3784e-04],
        [-2.4914e-04, -6.8649e-05,  3.3245e-04],
        [ 8.5931e-05, -7.7227e-05,  3.2653e-04],
        [-3.6088e-04, -9.9245e-05,  3.3297e-04],
        [ 7.3061e-04, -4.4464e-05,  3.8830e-04],
        [-7.7302e-05, -5.3050e-05,  3.8741e-04],
        [-1.8039e-04, -6.3788e-05,  3.8840e-04],
        [-3.3506e-04, -8.3120e-05,  3.9369e-04],
        [-1.8048e-04, -9.6546e-05,  3.9312e-04],
        [-1.5472e-04, -1.0192e-04,  3.9329e-04],
        [ 2.9230e-04, -7.1318e-05,  4.0359e-04],
        [-2.6643e-04, -7.6166e-05,  4.0543e-04]], grad_fn=<SelectBackward0>)
```
Typically, the shape of either were **(30, 3)** but the input required by my LSTM model is **(batch, 30, 3)**, that's why the unsqueeze on the 0th index. On training, its data's shape was **(32, 30, 3)** (I like it more when it's batch first, feats last).

### Solutions I've tried
1. Check the input shape and data.
2. Restart the kernel and run the jupyter lab once again.
3. Increase the Epoch.
4. Asked ChatGPT on what the cause could be, but the solution that bot gave ruined it more than fix it, like having more indicing problems, detaching issues to the point that I have to code back my original code and ask here instead.
5. I also inspected my data, but nothing is particularly interesting. 

### You might wonder: 
- As for why the dimension is (30, 3): that's just because the model needs to understand the previous patterns and so on. I've searched online about sequencing and found that not flattening them would produce better results specifically for the **Regression Problem** that I was doing.
- Whether I would flatten it to be (90) is something I am looking forwards to do, but I need direction first to not waste any more time.

**Comment by GANESH-NADKARNI** [comment]
Your learning rate is almost certainly the issue — `lr=0.1` for `AdamW` on an LSTM is very high (typical range is `1e-3` to `1e-4`). At that LR the optimizer overshoots on every step and the model settles into predicting a constant near the target's mean, since that's a cheap local minimum for the loss — this is a textbook "stuck output" pattern for RNNs/LSTMs.

Things to try, in order:

1. **Drop the LR** to `1e-3` first, then tune down/up from there:
optimizer = torch.optim.AdamW(params=model_0.parameters(), lr=1e-3)

2. **Train longer** — 3 epochs is very few for an LSTM to escape the trivial "predict the mean" solution even with a good LR. Try 50-100+ and watch if loss actually decreases per epoch (print it every epoch, not every 2).
3. Your features are all ~1e-4 magnitude — that's fine for NN input, but double check your **targets (`y_train`)** are a similar scale and not, say, raw prices. If targets and predictions are wildly different scales, the model can't learn anything useful even with the right LR.
4. Minor: you don't need `X_train.requires_grad_(True)` / `X_test.requires_grad_(True)` — inputs don't need gradients, only `model.parameters()` do. It's not the bug, but it's extra unnecessary autograd tracking on your data.

Try the LR change first — that's the single most likely fix.

**Comment by stefisha** [comment]
I don't think the input shape is the problem. `(batch, 30, 3)` is a valid shape for your LSTM since you're using `batch_first=True`, so I wouldn't flatten it to 90.

The first thing I would change is your learning rate:

```python
optimizer = torch.optim.AdamW(model_0.parameters(), lr=1e-3)
```

`lr=0.1` is very high for AdamW and can make the model quickly collapse to predicting almost the same value for every input.

I would also normalize/standardize both your input features and target values before training. Your inputs are very small (around `1e-4`), and scaling the data usually makes this type of regression much easier to train.

You also don't need:

```python
X_train.requires_grad_(True)
X_test.requires_grad_(True)
```

Gradients are needed for the model parameters, not your input data, so you can remove those lines.

I would start with a simpler model too:

```python
model_0 = ForexModelV0(
    input_size=3,
    hidden_size=32,
    output_size=1,
    layers=1
)

optimizer = torch.optim.AdamW(model_0.parameters(), lr=1e-3)
```

Then train it for more than 3 epochs and watch whether the loss is actually decreasing.

One useful test is to try to overfit a very small amount of data, for example one batch. If the model cannot learn one batch after many iterations, then there is likely still a problem with the training setup or data. If it can, then the model itself is probably fine and you can look at scaling/data quality next.

So I would try, in order: lower the learning rate, normalize the data, remove `requires_grad` from the inputs, use a 1-layer LSTM, and test whether it can overfit one batch.

If this resolves your issue, please mark it as the answer. 🙂
